File: Java/src/main/java/com/semantria/AutoResponseFeatureTestApp.java
package com.semantria;

import com.semantria.interfaces.ICallbackHandler;
import com.semantria.mapping.Document;
import com.semantria.mapping.configuration.Configuration;
import com.semantria.mapping.output.CollAnalyticData;
import com.semantria.mapping.output.DocAnalyticData;
import com.semantria.serializer.JsonSerializer;
import com.semantria.utils.RequestArgs;
import com.semantria.utils.ResponseArgs;

import java.io.FileReader;
import java.io.BufferedReader;
import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

import org.apache.commons.lang.StringUtils;

public class AutoResponseFeatureTestApp {
	
	public static void main(String [] args) throws InterruptedException 
	{
		final String consumerKey = "";
		final String consumerSecret = "";
				
		
		List<Document> docList = new ArrayList<Document>(100);
		final List<DocAnalyticData> resList = new ArrayList<DocAnalyticData>(100);
		
		System.out.println("Semantria Auto-response feature demo.");
		System.out.println();
		
		File file = new File( System.getProperty("user.dir") + "\\src\\main\\java\\" + DiscoveryModeTestApp.class.getPackage().getName().replace(".","\\") + "\\source.txt");
        
        if( !file.exists() )
        {
            System.out.println("Source file isn't available.");
            return;
        }
        
        //Reads collection from the source file
        BufferedReader br =null;
        
        try
        {
       
        	br = new BufferedReader(new FileReader(file));
            String strLine;

            while ((strLine = br.readLine()) != null)
            {	
            	
            	String uid = UUID.randomUUID().toString();
             
                if (StringUtils.isEmpty(strLine) || strLine.length() < 3)
                {
                    continue;
                }

                docList.add(new Document(uid, strLine));
            }
           
        }
        catch (Exception e){
            System.err.println("Error: " + e.getMessage());
        }finally {
        	
        	 if(br!=null)
				try {
					br.close();
				} catch (IOException e) {
					e.printStackTrace();
				}
        	
        }
        
        
        Session session = Session.createSession(consumerKey, consumerSecret, new JsonSerializer());
        session.setCallbackHandler(new ICallbackHandler() {
            @Override
            public void onResponse(Object sender, ResponseArgs responseArgs) {
            }

            @Override
            public void onRequest(Object sender, RequestArgs requestArgs) {
            }

            @Override
            public void onError(Object sender, ResponseArgs errorArgs) {
                System.out.println(String.format("%d: %s", (int) errorArgs.getStatus(), errorArgs.getMessage()));
            }

            @Override
            public void onDocsAutoResponse(Object sender, List<DocAnalyticData> processedData) {
            	
                for (DocAnalyticData data : processedData)
                {
                	resList.add(data);
                }
            	
            }

            @Override
            public void onCollsAutoResponse(Object sender, List<CollAnalyticData> processedData) {
            }
        });
        
        
     // Remembers primary configuration to set it back after the test.
        List<Configuration> configsList = session.getConfigurations();
        Configuration primaryConf = getPrimaryConfig(configsList);
        
       
        
     // Updates or Creates new configuration for the test purposes.
	    if(!compareName(configsList, "AutoResponseTest"))
	    {
	    	Configuration config = new Configuration();
	    	config.setName("AutoResponseTest");
	    	config.setIsPrimary(true);
	    	config.setLanguage("English");
	    	config.setAutoResponse(true);
	    	
	    	List<Configuration> newConfig = new ArrayList<Configuration>();
	    	newConfig.add(config);
	    	
	    	System.out.println(session.addConfigurations(newConfig));
	    	
	    	
	    }
	    else
	    {
	    	
	    	
	    	Configuration config = getConfigByName(configsList, "AutoResponseTest");
	    	config.setIsPrimary(true);
	    	List<Configuration> newConfig = new ArrayList<Configuration>();
	    	newConfig.add(config);
	    	session.updateConfigurations(newConfig);
	    	
	    	
	    }
	    
	    
	    
        // Queues documents for analysis one by one
        for (int i = 0; i < docList.size(); i++)
        {

            session.queueDocument(docList.get(i));
            Thread.sleep(100);

           System.out.println("Documents queued/received rate: " + (i + 1) + "/" + resList.size());
        }
        
        
        // The final call to get remained data from server, Just for demo purposes.
        Thread.sleep(1000);
        while (docList.size() != resList.size())
        {
            List<DocAnalyticData> lastResults = session.getProcessedDocuments();
            for(DocAnalyticData data : lastResults)
            {
                resList.add(data);
            }

            Thread.sleep(500);
        }
        
        System.out.println("Documents queued/received rate: " + docList.size()  + "/" +  resList.size());

        // Sets original primary configuration back after the test.
    	List<Configuration> newConfig = new ArrayList<Configuration>();
    	newConfig.add(primaryConf);
    	session.updateConfigurations(newConfig);
 			
	}
	
    public static boolean compareName(List<Configuration> config, String compare)
    {
    	boolean configExist = false;
    	for(Configuration configuration : config)
    	{
    		if(configuration.getName().equals(compare))
    		{
    			configExist = true;
    			break;
    			
    		}
    		
    	}
    	
    	return configExist;	
    }
    
    
    public static Configuration getPrimaryConfig(List<Configuration> config)
    {
    	
    	Configuration getConfig = null;
    	for(Configuration configuration : config)
    	{
    		if(configuration.getIsPrimary() == true)
    		{
    			getConfig =  configuration;
    			break;
    			
    		}
    			
    	}
    	
    	return getConfig;
    	
    }
    
    
    public static Configuration getConfigByName(List<Configuration> config, String name)
    {
    	
    	Configuration getConfig = null;
    	
    	for(Configuration configuration : config)
    	{
    		if(configuration.getName().equals(name))
    		{
    			getConfig =  configuration;
    			break;
    			
    		} 		
    		
    	}
    	   	
    	return getConfig;
    	
    }

}

File: PHP/examples/AutoResponseFeatureTestApp.php
<?php

require_once('../semantria/session.php');

echo "Semantria Auto-response feature demo.", "\r\n";

// the consumer key and secret
define('CONSUMER_KEY', "");
define('CONSUMER_SECRET', "");

$results = array();


/**
 * Simple Semantria callback's handler
 */
class SessionCallbackHandler extends \Semantria\CallbackHandler
{
    function onRequest($sender, $args)
    {
        //$s = json_encode($args);
        //echo "REQUEST: ", htmlspecialchars($s), "\r\n";
    }

    function onResponse($sender, $args)
    {
        //$s = json_encode($args);
        //echo "RESPONSE: ", htmlspecialchars($s), "\r\n";
    }

    function onError($sender, $args)
    {
        $s = json_encode($args);
        echo "ERROR: ", htmlspecialchars($s), "\r\n";
    }

    function onDocsAutoResponse($sender, $response)
    {
        global $results;

        foreach ($response as $item) {
            $results[] = $item;
        }
    }

    function onCollsAutoResponse($sender, $args)
    {
        //$s = json_encode($args);
        //echo "COLLS AUTORESPONSE: ", htmlspecialchars($s), "\r\n";
    }
}

print("Reading collection from file...\n");
$documents = array();
$file = fopen("source.txt", "r");
if (!$file) {
    print("File not found.");
    exit(1);
}

while (($line = fgets($file)) !== FALSE) {
    $documents[] = $line;
}
if (!feof($file)) {
    echo "Error: unexpected fgets() fail\n";
}
fclose($file);

$total_docs = count($documents);

// Initializes new session with the serializer object and the keys.
$session = new \Semantria\Session(CONSUMER_KEY, CONSUMER_SECRET, NULL, NULL, TRUE);

// Initialize session callback handler
$callback = new SessionCallbackHandler();
$session->setCallbackHandler($callback);

$configurations = $session->getConfigurations();
$primary_config = NULL;
$autoresponse_config = NULL;

foreach ($configurations as $configuration) {
    if ($configuration['is_primary']) {
        $primary_config = $configuration;
    } 
    else if ($configuration['name'] == 'AutoResponseTest') {
        $autoresponse_config = $configuration;
    }
}

if ($autoresponse_config == NULL) {
    $session->addConfigurations(array(array(
        'name' => 'AutoResponseTest',
        'language' => 'English',
        'is_primary' => TRUE,
        'auto_response' => TRUE,
    )));
}
else {
    $autoresponse_config['is_primary'] = TRUE;
    $session->updateConfigurations(array($autoresponse_config));
}

// Queues documents for analysis one by one
$doc_counter = 0;
foreach ($documents as $doc) {
    $session->queueDocument(array('id' => uniqid('', TRUE), 'text' => $doc));
    $doc_counter += 1;
    usleep(100000);

    $results_len = count($results);
    print("Documents queued/received rate: ${doc_counter}/${results_len}\n");
}

// The final call to get remained data from server, Just for demo purposes.
sleep(1);
while (count($results) < count($documents)) {
    $result = $session->getProcessedDocuments();

    foreach ($result as $item) {
        $results[] = $item;
    }

    usleep(500000);
}

$results_len = count($results);
print("Documents queued/received rate: ${total_docs}/${results_len}\n");

// Sets original primary configuration back after the test.
$session->updateConfigurations(array($primary_config));
